state/state.go

```diff
@@ -34,6 +40,110 @@ type DownloadedObject struct {
 	Object, Filename string
 }
 
+type DynamoDBStater struct {
+	*sync.Mutex
+	Session *session.Session
+	Service string
+}
+
+func NewDynamoDBStater(session *session.Session, service string) *DynamoDBStater {
+	return &DynamoDBStater{
+		Mutex:   &sync.Mutex{},
+		Session: session,
+		Service: service,
+	}
+}
+
+// list of processed objects
+func (d *DynamoDBStater) processedObjects() (map[string]time.Time, error) {
+	objs := make(map[string]time.Time)
+
+	type Record struct {
+		S3object string    `json:"S3object"`
+		Time     time.Time `json:"Time"`
+	}
+
+	var records []Record
+
+	svc := dynamodb.New(d.Session)
+	err := svc.ScanPages(&dynamodb.ScanInput{
+		TableName: aws.String("HoneyELBAccessLogBuckets"),
+	}, func(logs *dynamodb.ScanOutput, last bool) bool {
+		recs := []Record{}
+
+		err := dynamodbattribute.UnmarshalListOfMaps(logs.Items, &recs)
+
+		// break out of function
+		if err != nil {
+			panic(fmt.Sprintf("failed to unmarshal Dynamodb Scan Items, %v", err))
+		}
+
+		records = append(records, recs...)
+
+		return true
+	})
+
+	if err != nil {
+		return objs, fmt.Errorf("Error scanning DynamoDB, %v", err)
+	}
+
+	for _, record := range records {
+		objs[record.S3object] = record.Time
+	}
+	return objs, nil
+}
+
+func (d *DynamoDBStater) ProcessedObjects() (map[string]time.Time, error) {
+	d.Lock()
+	defer d.Unlock()
```

Comment:
What are we protecting with this lock?

Reply:
I think in DynamoDB's case it will not add much, since we are essentially off-shoring the consistency logic to Amazon. It is used in FileStater as a "just in case" for access across goroutines. @snoozan do you think this can be removed?